Frameworks/Foundation/NSDecimal.mm

```diff
@@ -0,0 +1,605 @@
+}
+
+static NSCalculationError _compactMantissa(NSDecimalInternal* iDecimal, NSRoundingMode roundingMode) {
+    // Note: All inputs have been normalized via NSDecimalNormalize
+    // divide by 10 until we can compact it.
+    int exponent = 0;
+
+    NSDecimalInternal result;
+    while (iDecimal->decimal._length > NSDecimalMaxSize) {
+        exponent++;
+        NSCalculationError error = _mantissaDivision(iDecimal, iDecimal, 10);
+        if (error != NSCalculationNoError) {
+            return error;
+        }
+    }
+
+    // update the exp
+    result.decimal._exponent += exponent;
+
+    // TODO: Add Support for rounding
+    if (result._remainder != 0) {
+    }
+
+    return NSCalculationNoError;
+}
+
+/**
+ @Status Stub
+ @Notes
+*/
+NSString* NSDecimalString(const NSDecimal* dcm, id locale) STUB_METHOD {
+    UNIMPLEMENTED();
+    return StubReturn();
+}
+
+/**
+ @Status Interoperable
+*/
+NSComparisonResult NSDecimalCompare(const NSDecimal* leftOperand, const NSDecimal* rightOperand) {
+    if (leftOperand == rightOperand) {
+        return NSOrderedSame;
+    }
+
+    THROW_NS_IF_FALSE(E_INVALIDARG, (leftOperand != NULL) && (rightOperand != NULL));
+
+    NSDecimal left, right;
+    NSDecimalCopy(&left, leftOperand);
+    NSDecimalCopy(&right, rightOperand);
+    NSDecimalNormalize(&left, &right, 0);
+
+    if (!left._isNegative && right._isNegative) {
+        return NSOrderedDescending;
+    }
+
+    if (left._isNegative && !right._isNegative) {
+        return NSOrderedAscending;
+    }
+
+    // same signs
+    if (left._length < right._length) {
+        return right._isNegative ? NSOrderedDescending : NSOrderedAscending;
+    }
+
+    if (left._length > right._length) {
+        return right._isNegative ? NSOrderedAscending : NSOrderedDescending;
+    }
+
+    // all same
+    int index = right._length - 1;
+    while (index >= 0) {
+        if (left._mantissa[index] < right._mantissa[index]) {
+            return right._isNegative ? NSOrderedDescending : NSOrderedAscending;
+        }
+
+        if (left._mantissa[index] > right._mantissa[index]) {
+            return right._isNegative ? NSOrderedAscending : NSOrderedDescending;
+        }
+        index--;
+    }
+
+    return NSOrderedSame;
+}
+
+/**
+ @Status Caveat
+ @Notes roundingMode is not supported.
+*/
+NSCalculationError NSDecimalNormalize(NSDecimal* number1, NSDecimal* number2, NSRoundingMode roundingMode) {
+    THROW_NS_IF_FALSE(E_INVALIDARG,
+                      (number1 != NULL) && (number2 != NULL) && !NSDecimalIsNotANumber(number1) && !NSDecimalIsNotANumber(number2));
+
+    if (number1->_exponent == number2->_exponent) {
+        return NSCalculationNoError;
+    }
+
+    NSDecimal* decimalToNormalize;
+    NSDecimal* refDecimal;
+
+    if (number1->_exponent > number2->_exponent) {
+        decimalToNormalize = number1;
+        refDecimal = number2;
+    } else {
+        decimalToNormalize = number2;
+        refDecimal = number1;
+    }
```

Comment:
good point. i'm going to open a bug for this, as it would need some more testing after the change.

---
In reply to: [81217804](https://github.com/Microsoft/WinObjC/pull/1032#discussion_r81217804) [](ancestors = 81217804,80622059,80588035)